rfcs/2020-05-25-more-usable-logevents.md

```diff
@@ -0,0 +1,223 @@
+    pub fn remove(&mut self, key: &Atom) -> Option<Value>;
+    pub fn remove_prune(&mut self, key: &Atom, prune: bool) -> Option<Value>;
+    pub fn keys<'a>(&'a self) -> impl Iterator<Item = String> + 'a;
+    pub fn all_fields<'a>(&'a self) -> impl Iterator<
+				Item = (String, &'a Value)
+		> + Serialize;
+    pub fn is_empty(&self) -> bool;
+}
+```
+
+Many of these APIs offer functionality ***coarsely similar*** to a `FlatMap`. These APIs were added ad-hoc, as needed over time, and were not designed intentionally to all fit together. In addition, some changes were added unobtrusively to aid in the migration path.
+
+### Example `LogEvent` Lifetimes
+
+- **Simple Connector:** A `LogEvent` entering Vector through a `http(encoding=json)//console(encoding=text)` pipeline.
+- **Runtime Pipeline:** A `LogEvent` entering Vector through a `http(encoding=json)/lua/http(encoding=json,buffer=disk)` pipeline.
+- **N:M:O sources, transforms, and sinks:** A `LogEvent` entering Vector through a `???/????/???` pipeline.
+
+# Guide Level Proposal
+
+There is no guide accompanying this RFC, it only minimally touches user facing surface.
+
+# Doc Level Proposal
+
+Insert into [Log Event](https://vector.dev/docs/about/data-model/log/#types)'s [Types](https://vector.dev/docs/about/data-model/log/#types) section
+
+### Bytes
+
+An arbitrary sequence of bytes (not necessarily UTF-8), bounded by system memory. If Vector is unable to parse a String as UTF-8 bytes, it will use this instead of mutating the String to make it UTF-8.
+
+Modify the [Coercer transform](https://vector.dev/docs/reference/transforms/coercer/#field-name) Types section.
+
+```bash
+Enum, must be one of: "bool" "float" "int" "string" "timestamp", "bytes"
+```
+
+# Prior Art
+
+In the [WASM transform](https://github.com/timberio/vector/pull/2006/files) our demo POC was using protobufs, which by their nature are not always UTF-8. We had to patch Vector to support Bytes during this time. It currently does so via a fallback when non-UTF-8 bytes are found.
+
+# Sales Pitch
+
+This RFC ultimately proposes the following steps:
+
+1. Rename `Value::Bytes` to `Value::String` .
+2. Add `Value::Bytes`.
+3. Significant UX improvements on `LogEvent`, particularly turning JSON ↔ `LogEvent`.
+4. Refactor the `PathIter` to make `vector::Event::Path` type.
+5. Significant UX improvements on `Path` , particularly an internal `String` ↔ `Path` with an `Into` that does not do path parsing, as well as a `Path::new(s: String)` that does.
+6. Refactor all `LogEvent` to accept `Into<Path>` values.
+    1. Remove obsolete functionality like `insert_path` and `remove_prune` since the new `Path` type covers this.
+    2. Refactor the `keys` function to return an `Iterator<Path>`
+7. Add an `Entry` style API to `LogEvent`.
+    1. Remove functionality rendered obsolete by the Entry API like `try_insert`, moving them to use the new Entry API
+8. Provide `iter` and `iter_mut` functions that return `(Path, Value)`. 
```

Comment:
Nit pick: these functions won't return that type, they will yield it (or yield elements of type …).